### PR TITLE
Added wikifier to the preprocessing package

### PR DESCRIPTION
**Pull request by deniselezi**
Created a Wikifier class in the preprocessing package. A potential user would import this class and instantiate it with their Wikifier API key. They would then be able to make to make a request to the Wikifier API using the wikify method.

### CODE REVIEW

File: test.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+from truelearn.preprocessing.wikifier import Wikifier
```
[the remark this answers is not in the document]
Comment (deniselezi) [reply test.py]: I was just running test.py from time to time to ensure that I didn't break Sahan's code's functionality by refactoring it. Following your feedback regarding credentials.txt, I have moved both it and test.py to the parent folder.

**Comment by deniselezi** [comment]
todo:

- [ ] Add a reference to Wikifier in the code (probably through a docstring) as requested in the Wikifier documentation
- [ ] Raise more relevant exceptions instead of ValueError
- [ ] Remove temporary benchmark file

File: requirements.txt [hunk 2]
```diff
@@ -1,4 +1,5 @@
 numpy
```
Comment (yuxqiu) [inline requirements.txt:1]: You can revert the changes in `requirements.txt` and `setup.py` as they are not used in the main branch.

File: truelearn/preprocessing/wikifier.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
+import orjson
+from urllib import request, parse, error
+
+from typing import Optional, Union, NoReturn
+Annotation = dict[str, Union[str, float, None]]
+WikifierResponse = dict[str, Union[list[str], list[dict]]]
+
+
+class Wikifier:
+    """
```
Comment (yuxqiu) [inline truelearn/preprocessing/wikifier.py:10]: We use a different style for docstring now. Please refer to [Google Python Style Guide](https://google.github.io/styleguide/pyguide.html).

For most of your code, you need to use the `Args`, `Returns` and `Raises` specified in the guide.

File: truelearn/preprocessing/wikifier.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
+import orjson
```
Comment (yuxqiu) [inline truelearn/preprocessing/wikifier.py:1]: The order of import should comform to google python style guide.

File: truelearn/preprocessing/wikifier.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    def __init__(self, api_key: str) -> None:
+        if isinstance(api_key, str):
+            self.api_key = api_key
+        else:
+            raise TypeError("api_key should be a string")   
+
+    def wikify(
+            self, text: str, df_ignore: int = 50,
+            words_ignore: int = 50, top_n: Optional[int] = None
+        ) -> Union[list[Annotation], NoReturn]:
```
Comment (yuxqiu) [inline truelearn/preprocessing/wikifier.py:42]: I suspect that the use of NoReturn is incorrect. [NoReturn vs. None in "void" functions - type annotations in Python 3.6](https://stackoverflow.com/questions/48038149/noreturn-vs-none-in-void-functions-type-annotations-in-python-3-6)

File: truelearn/preprocessing/wikifier.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+                text, df_ignore, words_ignore
+            )
+        except error.HTTPError as err:
+            raise RuntimeError(str(err)) from err
+
+        return self.__format_wikifier_response(resp, top_n)
+
+    def __make_wikifier_request(
+            self, text: str, df_ignore: int, words_ignore: int
+        ) -> Union[WikifierResponse, NoReturn]:
```
Comment (yuxqiu) [inline truelearn/preprocessing/wikifier.py:86]: Notice the `NoReturn` here.

File: requirements.txt (hunk 2 above)
Comment (deniselezi) [reply requirements.txt:1]: Do you mean that I should revert them to the last commit before I opened this branch?

File: truelearn/preprocessing/wikifier.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            RuntimeError: The HTTP request returns a status code representing
+              an error
+        """
+        try:
+            resp = self.__make_wikifier_request(
+                text, df_ignore, words_ignore
+            )
+        except error.HTTPError as err:
+            raise RuntimeError(str(err)) from err
+
```
Comment (deniselezi) [inline truelearn/preprocessing/wikifier.py:63]: @yuxqiu @KD-7 do you have any thoughts on this? I'm raising a RuntimeError from urllib's HTTPError so that the user does not have to import urllib to handle it, which seems like it would be a bad practice. Is there a better way to do this?

**Comment by KD-7** [comment]
Necessary features to add as discussed:
- [ ] Mean of cosine similarities 
- [ ] Standard deviation of cosine similarities

**Comment by KD-7** [comment]
Ive just resolved the merge conflicts

File: requirements.txt (hunk 2 above)
Comment (KD-7) [reply requirements.txt]: Resolved in my commits

File: truelearn/preprocessing/wikifier.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        """
+        params = {
+            'text': text,
+            'userKey': self.api_key,
+            'nTopDfValuesToIgnore': df_ignore,
+            'nWordsToIgnoreFromList': words_ignore
+        }
+
+        data = parse.urlencode(params)
+        url = "http://www.wikifier.org/annotate-article?" + data
```
Comment (KD-7) [inline truelearn/preprocessing/wikifier.py:86]: make it https

File: truelearn/preprocessing/wikifier.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+    .. [1] Janez Brank, Gregor Leban, Marko Grobelnik. Annotating Documents
+       with Relevant Wikipedia Concepts. Proceedings of the Slovenian Conference
+       on Data Mining and Data Warehouses (SiKDD 2017), Ljubljana, Slovenia,
+       9 October 2017.
+    """
+
+    def __init__(self, api_key: str) -> None:
+        """Inits Wikifier class with api_key."""
+        if isinstance(api_key, str):
```
Comment (KD-7) [inline truelearn/preprocessing/wikifier.py:27]: Im wondering if we need this because of the type hint we enforced

File: truelearn/preprocessing/wikifier.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+              or the API key is not valid.
+        """
+        params = {
+            'text': text,
+            'userKey': self.api_key,
+            'nTopDfValuesToIgnore': df_ignore,
+            'nWordsToIgnoreFromList': words_ignore
+        }
+
+        data = parse.urlencode(params)
```
Comment (KD-7) [inline truelearn/preprocessing/wikifier.py:85]: for data dont you need to encode this e.g. utf-8?

**Review by KD-7 — COMMENTED** [review]
Please see my comments. Also you havent modified the __init__.py in preprocessing, so please add the doc string and __all__ = ["Wikifier"] etc

File: truelearn/preprocessing/wikifier.py (hunk 7 above)
Comment (KD-7) [reply truelearn/preprocessing/wikifier.py:64]: You marked it as resolved, you could handle the 404 error if you wanted to

Comment (deniselezi) [reply truelearn/preprocessing/wikifier.py:64]: I'm handling the error by chaining it into a different one that the user can deal with, I don't know what else I would do to handle it, besides perhaps adding more information to the Runtime error before raising it

Comment (KD-7) [reply truelearn/preprocessing/wikifier.py:64]: > I'm handling the error by chaining it into a different one that the user can deal with, I don't know what else I would do to handle it, besides perhaps adding more information to the Runtime error before raising it

see my updated comment

Comment (deniselezi) [reply truelearn/preprocessing/wikifier.py:64]: You mean except the HTTPError in the __make_wikifier_request method rather than in wikify? Will do

File: truelearn/preprocessing/wikifier.py (hunk 10 above)
Comment (deniselezi) [reply truelearn/preprocessing/wikifier.py:110]: I'm not entirely sure, but the response is correct and it looks like the encoding is carried out by the urlencode method.

File: truelearn/preprocessing/wikifier.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        url = "http://www.wikifier.org/annotate-article?" + data
+        with request.urlopen(url) as r:
+            resp = r.read()
+            resp = orjson.loads(resp)
+            if 'error' in resp:
+                raise ValueError(f"error in response : {resp['error']}")
+            if 'status' in resp:
+                # will trigger if key is not valid
+                raise ValueError(resp['status'])
+            return resp
```
[the remark this answers is not in the document]
Comment (KD-7) [reply truelearn/preprocessing/wikifier.py:120]: add a comment stating this, that it must be utf -8 and maybe with the link

File: truelearn/preprocessing/wikifier.py [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+        with request.urlopen(url) as r:  # nosec
+            resp = orjson.loads(r.read().decode("utf-8"))
+            if "error" in resp:
+                raise ValueError(f"error in response : {resp['error']}")
+            if "status" in resp:
+                # will trigger if key is not valid
+                raise ValueError(resp["status"])
+            return resp
+
+    def __format_wikifier_response(
```
Comment (KD-7) [inline truelearn/preprocessing/wikifier.py:146]: format wikifier response may be static warning

**Review by KD-7 — CHANGES_REQUESTED** [review]
For the file names to be consistent we append _ to the start of the filename?

**Comment by KD-7** [comment]
Additionally, @deniselezi needs to confirm whether this standard deviation and mean calculation is what was meant by @sahanbull

**Comment by sahanbull** [comment]
> 

yes. these are correct. We need sample mean and standard deviation of cosine values.
